### ISSUE TITLE
speech.stop() infinitive blocking - Android 10

### ISSUE THREAD

**Issue by nick-uk** [csdcorp/speech_to_text#146]
Hi,

Great plugin but I realised I have a potential bug:

```dart
speech.listen(
    onResult: resultListener,
    listenFor: Duration(seconds: 10),
    localeId: _currentLocaleId,
    onSoundLevelChange: soundLevelListener,
    cancelOnError: true,
    listenMode: ListenMode.deviceDefault, // same behaviour with ListenMode.confirmation
 );

print('[' + DateTime.now().toString() + '] -> DEBUG: stopListening() start');
await speech.stop(); // <--- Try to stop it after 5 seconds but blocks forever
print('[' + DateTime.now().toString() + '] -> DEBUG: stopListening() stop');
```

> **Version: speech_to_text-2.7.0**
> Flutter (Channel stable, 1.22.2, on Mac OS X 10.15.7 19H2, locale en-GB)
> Flutter version 1.22.2 
> Dart version 2.10.2
> Android toolchain - develop for Android devices (Android SDK version 29.0.3)
> Platform android-30, build-tools 29.0.3
> minSdkVersion 21
> targetSdkVersion 29

It appears on Android OS speech.stop() never returns. On IOS works as expected. After several speech.listen(), speech.stop() circles the app crashes without a lot of information so the behaviour of the speech.stop() is my main suspect.

I would appreciate some feedback.

Thanks,
Nick

**Comment by sowens-csd** [comment]
Sorry Nick, not sure how I missed this issue. In the code snippet you showed it looks like the `speech.stop` call happens immediately after the call to `speech.listen` without an `await` on the listen. While there's nothing technically forbidding doing that I'm not sure that I've tested that behaviour. Can you confirm that's what you're doing? Can you try it with an `await` on the listen if there isn't one?

**Comment by nick-uk** [comment]
Hi,

Thanks for your answer.
The snippet was a short version of what I'm trying to do. I can reproduce it like this as well:

```dart
speech.listen(
    onResult: resultListener,
    listenFor: Duration(seconds: 10),
    localeId: _currentLocaleId,
    onSoundLevelChange: soundLevelListener,
    cancelOnError: true,
    listenMode: ListenMode.deviceDefault, // same behaviour with ListenMode.confirmation
 );

await Future.delayed(Duration(seconds: 5));
print('[' + DateTime.now().toString() + '] -> DEBUG: stopListening() start');
await speech.stop(); // <--- Try to stop it after 5 seconds but blocks forever
print('[' + DateTime.now().toString() + '] -> DEBUG: stopListening() stop');
```

output:
```log
[2020-11-09 08:55:00.300659] -> DEBUG: speech.listen() start
[2020-11-09 08:55:05.306112] -> DEBUG: speech.stop() start
# blocks here
```
If I will add an await in front of speech.listen() and call it from within an async function nothing change. Last time I tested it on IOS the speech.stop() returns as expected.

I hope this helps,
Regards,

Nick G

**Comment by sowens-csd** [comment]
Are you reproducing this on a simulator or real device? What Android version?

**Comment by nick-uk** [comment]
Hi Sowens, 

I just re-test it on a real device with better logging to reproduce and it's tricky. It seems is working only if I have at least one recognized word in previous speech.listen() circle. When the microphone remains silent, in the case of the emulator where I don't have input at all, or when I'm not speaking when it listens, the speech.stop() blocks forever. Both emulator and real are Android 10.

For example:

speech.listen() <- Here there is voice input
speech.stop()  <- This is not returning
speech.listen() <- With ot without voice input
speech.stop()  <- Now returns
---
speech.listen() <- Here there is no voice input
speech.stop()  <- This is not returning
speech.listen() <- Here there is voice input
speech.stop()  <- This is not returning

I will try to allocate some time to do more tests.

I hope this helps,
Regards,

Nick G

**Comment by sowens-csd** [comment]
This is helpful, thanks. I'll see if I can reproduce.

**Comment by sowens-csd** [comment]
I was able to reproduce the behaviour on an emulator. I think it is fixed with my latest commit. If you have a chance and can try it before I release that would be helpful. Thanks for the report. 

Turns out that `stop` and other methods would never return a result if the plugin wasn't actively listening at the time they were called. I think what was happening is that either the speech timeout or an error was making the `listen` call fail so the `stop` was being called when not listening. That should be perfectly legal and just do nothing but the bug meant that it did nothing but never returned a result to the caller.

**Comment by nick-uk** [comment]
Hi Sowens,

I test your plugin from master branch (last commit c57320410a087f98c8edc0f5abf3aadc6a981416) and now **speech.stop()** is working as expected for both **IOS 14.1** and **Android 10** as expected! Master includes a fix for both #146 and #138 issues. I don't have any issues with voice recognition with IOS 14.1 on iPhone 12 Pro simulator or the speech.stop() issue.

My configuration is:

>    • Flutter version 1.22.2 at /Users/nick/flutter
>    • Dart version 2.10.2
>
>    • Platform android-30, build-tools 29.0.3
>
>   • Xcode 12.1, Build version 12A7403
>   • CocoaPods version 1.8.4
>
>  Connected device (2 available)                     
>     •  Android 10 (API 29) (emulator)
>     • iPhone 12 Pro (mobile), com.apple.CoreSimulator.SimRuntime.iOS-14-1 (simulator)

Thanks a lot for the fix. I will do more tests if I have any new issues I will let you know.

Regards,
Nick G
